src/sage/quadratic_forms/bqf_class_group.py

```diff
@@ -0,0 +1,608 @@
+            sage: cl = Cl.zero(); cl
+            Class of x^2 + x*y + 250*y^2
+            sage: cl + cl == cl
+            True
+        """
+        return self(BinaryQF.principal(self._disc))
+
+    def random_element(self):
+        r"""
+        Return a somewhat random element of this form class group.
+
+        ALGORITHM:
+
+        Sample random odd primes `a` until `b^2 = D \pmod{4a}` has a
+        solution `b \in \ZZ` and set `c = (b^2-D)/(4a)`. Flip a coin
+        to choose the sign of `b`. Then return the class of `[a,b,c]`.
+
+        .. NOTE::
+
+            No strict guarantees are being made about the distribution of
+            classes sampled by this function. Heuristically, however, it
+            should be fairly close to uniform.
+
+        EXAMPLES::
+
+            sage: Cl = BQFClassGroup(-999); Cl
+            Form Class Group of Discriminant -999
+            sage: cl = Cl.random_element(); cl  # random
+            Class of 10*x^2 + x*y + 25*y^2
+            sage: cl.form().discriminant()
+            -999
+        """
+        B = self._disc.abs() * 100 + 9999
+        while True:
+            a = random_prime(B, proof=False, lbound=3)
+            if self._disc.kronecker(a) == 1:
+                break
+        b = ZZ(Mod(self._disc, 4*a).sqrt())
+        c = (b**2 - self._disc) // (4*a)
+        if randrange(2):
+            b = -b
+        return self(BinaryQF([a,b,c]))
+
+    def __hash__(self):
+        r"""
+        Return a hash value for this form class group.
+
+        EXAMPLES::
+
+            sage: hash(BQFClassGroup(-999))  # random
+            -4246560339810542104
+        """
+        return hash(('BQFClassGroup', self._disc))
+
+    def _repr_(self):
+        r"""
+        Return a string describing this form class group.
+
+        EXAMPLES::
+
+            sage: BQFClassGroup(-999)  # indirect doctest
+            Form Class Group of Discriminant -999
+        """
+        return f'Form Class Group of Discriminant {self._disc}'
+
+    def discriminant(self):
+        r"""
+        Return the discriminant of the forms in this form class group.
+
+        EXAMPLES::
+
+            sage: BQFClassGroup(-999).discriminant()
+            -999
+        """
+        return self._disc
+
+    @cached_method
+    def order(self):
+        r"""
+        Return the order of this form class group (the *class number*).
+
+        ALGORITHM: :pari:`qfbclassno`
+
+        EXAMPLES::
+
+            sage: BQFClassGroup(-4).order()
+            1
+            sage: BQFClassGroup(-11).order()
+            1
+            sage: BQFClassGroup(-67).order()
+            1
+            sage: BQFClassGroup(-163).order()
+            1
+            sage: BQFClassGroup(-999).order()
+            24
+            sage: BQFClassGroup(-9999).order()
+            88
+            sage: BQFClassGroup(-99999).order()
+            224
+        """
+        return ZZ(pari.qfbclassno(self._disc))
```

Comment:
When the discriminant is large, pari has `quadclassunit()` which has subexp. complexity (the highlighted function is just BSGS). (https://pari.math.u-bordeaux.fr/dochtml/html/Arithmetic_functions.html)

Might be worth always using the subexp and having a slow down for smaller disc, or having an if / else based on the size. Here pari suggests using it for D < 10^25, but with all the bindings etc this cut off point may vary

Reply:
Oh! I see you already use this in the `abelian_group()` method, so maybe indeed this could do the computation with this faster method for all cases.

Reply:
Thanks, I gave it a shot: Indeed, quadratic orders were also always using `qfbclassno()`, so I've added a unified wrapper to call one or the other depending on the discriminant.

Reply:
Nice! The wrapper seems like it helped more generally too, so that's good. 🥳